File: app/src/main/java/dev/ragnarok/fenrir/util/refresh/RefreshToken.java
package dev.ragnarok.fenrir.util.refresh;

import static dev.ragnarok.fenrir.util.Utils.isEmpty;

import android.util.Log;

import java.util.List;

import dev.ragnarok.fenrir.AccountType;
import dev.ragnarok.fenrir.Constants;
import dev.ragnarok.fenrir.Injection;
import dev.ragnarok.fenrir.settings.Settings;
import dev.ragnarok.fenrir.util.Utils;
import io.reactivex.rxjava3.core.Single;


public class RefreshToken {
    private static boolean upgradeTokenKate(int account, String oldToken) {
        if (Utils.isHiddenAccount(account)) {
            return false;
        }
        String gms = TokenModKate.requestToken();
        if (gms == null) {
            return false;
        }
        String token = Injection.provideNetworkInterfaces().vkDefault(account).account().refreshToken(gms, null, null, null).blockingGet().token;
        Log.w("refresh", oldToken + " " + token + " " + gms);
        if (oldToken.equals(token) || isEmpty(token)) {
            return false;
        }
        Settings.get().accounts().storeAccessToken(account, token);
        return true;
    }

    private static boolean upgradeTokenOfficial(int account, String oldToken) {
        if (Utils.isHiddenAccount(account)) {
            return false;
        }
        List<String> gms = TokenModOfficialVK.requestToken();
        if (gms == null) {
            return false;
        }
        long timestamp = System.currentTimeMillis();
        String token = Injection.provideNetworkInterfaces().vkDefault(account).account().refreshToken(gms.get(0), gms.get(1), TokenModOfficialVK.getNonce(timestamp), timestamp).blockingGet().token;
        Log.w("refresh", oldToken + " " + token + " " + gms);
        if (oldToken.equals(token) || isEmpty(token)) {
            return false;
        }
        Settings.get().accounts().storeAccessToken(account, token);
        return true;
    }

    public static boolean upgradeToken(int account, String oldToken) {
        if (Constants.DEFAULT_ACCOUNT_TYPE == AccountType.KATE) {
            return upgradeTokenKate(account, oldToken);
        } else if (Constants.DEFAULT_ACCOUNT_TYPE == AccountType.VK_ANDROID) {
            return upgradeTokenOfficial(account, oldToken);
        }
        return false;
    }

    public static Single<Boolean> upgradeTokenRx(int account, String oldToken) {
        return Single.create(v -> {
            try {
                upgradeToken(account, oldToken);
                v.onSuccess(true);
            } catch (Exception ignored) {
                v.onSuccess(false);
            }
        });
    }
}

File: app/src/main/kotlin/dev/ragnarok/fenrir/util/refresh/RefreshToken.kt
package dev.ragnarok.fenrir.util.refresh

import android.util.Log
import dev.ragnarok.fenrir.AccountType
import dev.ragnarok.fenrir.Constants
import dev.ragnarok.fenrir.Includes.networkInterfaces
import dev.ragnarok.fenrir.settings.Settings
import dev.ragnarok.fenrir.util.Utils
import io.reactivex.rxjava3.core.Single

object RefreshToken {
    /*
    private fun upgradeTokenKate(account: Int, oldToken: String): Boolean {
        if (Utils.isHiddenAccount(account)) {
            return false
        }
        val ret = Single.create<Boolean> { s ->
            FirebaseInstallations.getInstance().getToken(true).addOnSuccessListener {
                val token =
                    networkInterfaces.vkDefault(account).account()
                        .refreshToken(it.token, null, null, null)
                        .blockingGet().token
                if (oldToken == token || token.isNullOrEmpty()) {
                    s.onSuccess(false)
                } else {
                    Settings.get().accounts().storeAccessToken(account, token)
                    s.onSuccess(true)
                }
            }.addOnFailureListener { s.onSuccess(false) }
        }.blockingGet()

        return ret
    }
     */

    private fun upgradeTokenKate(account: Int, oldToken: String): Boolean {
        if (Utils.isHiddenAccount(account)) {
            return false
        }
        val gms = TokenModKate.requestToken() ?: return false
        val token =
            networkInterfaces.vkDefault(account).account().refreshToken(gms, null, null, null)
                .blockingGet().token
        Log.w("refresh", "$oldToken $token $gms")
        if (oldToken == token || token.isNullOrEmpty()) {
            return false
        }
        Settings.get().accounts().storeAccessToken(account, token)
        return true
    }

    private fun upgradeTokenOfficial(account: Int, oldToken: String): Boolean {
        if (Utils.isHiddenAccount(account)) {
            return false
        }
        val gms = TokenModOfficialVK.requestToken() ?: return false
        val timestamp = System.currentTimeMillis()
        val token = networkInterfaces.vkDefault(account).account()
            .refreshToken(gms[0], gms[1], TokenModOfficialVK.getNonce(timestamp), timestamp)
            .blockingGet().token
        Log.w("refresh", "$oldToken $token $gms")
        if (oldToken == token || token.isNullOrEmpty()) {
            return false
        }
        Settings.get().accounts().storeAccessToken(account, token)
        return true
    }

    fun upgradeToken(account: Int, oldToken: String): Boolean {
        if (Constants.DEFAULT_ACCOUNT_TYPE == AccountType.KATE) {
            return upgradeTokenKate(account, oldToken)
        } else if (Constants.DEFAULT_ACCOUNT_TYPE == AccountType.VK_ANDROID) {
            return upgradeTokenOfficial(account, oldToken)
        }
        return false
    }

    fun upgradeTokenRx(account: Int, oldToken: String): Single<Boolean> {
        return Single.create {
            try {
                upgradeToken(account, oldToken)
                it.onSuccess(true)
            } catch (ignored: Exception) {
                it.onSuccess(false)
            }
        }
    }
}